server/src/main/java/io/crate/expression/operator/all/AllEqOperator.java

```diff
@@ -0,0 +1,122 @@
+import io.crate.sql.tree.ComparisonExpression;
+import io.crate.types.TypeSignature;
+
+public class AllEqOperator extends AllOperator {
+
+    public static String NAME = OPERATOR_PREFIX + ComparisonExpression.Type.EQUAL.getValue();
+    public static Signature SIGNATURE = Signature.builder(AllEqOperator.NAME, FunctionType.SCALAR)
+            .argumentTypes(TypeSignature.parse("E"), TypeSignature.parse("array(E)"))
+            .returnType(Operator.RETURN_TYPE.getTypeSignature())
+            .typeVariableConstraints(typeVariable("E"))
+            .features(Feature.DETERMINISTIC, Feature.STRICTNULL)
+            .build();
+
+    public AllEqOperator(Signature signature, BoundSignature boundSignature, IntPredicate cmp) {
+        super(signature, boundSignature, cmp);
+    }
+
+    public static Query refMatchesAllArrayLiteral(Reference probe, List<?> values, LuceneQueryBuilder.Context context) {
+        // col = ALL ([1,2,3]) --> col=1 and col=2 and col=3
+        LinkedHashSet<?> uniqueValues = new LinkedHashSet<>(values);
+        if (uniqueValues.contains(null)) {
+            return new MatchNoDocsQuery("Cannot match nulls");
+        }
+        if (uniqueValues.isEmpty()) {
+            return new MatchNoDocsQuery("Cannot match unless there is at least one non-null candidate");
+        }
+        if (uniqueValues.size() > 1) {
+            // `col=x and col=y` evaluates to `false` unless `x == y`
+            return new MatchNoDocsQuery("A single value cannot match more than one unique values");
+        }
+        var value = uniqueValues.getFirst();
+        // col = all([1]) --> col = 1
+        Function eq = new Function(
+            EqOperator.SIGNATURE,
+            List.of(probe, Literal.ofUnchecked(probe.valueType(), value)),
+            EqOperator.RETURN_TYPE
+        );
+        return context.visitor().visitFunction(eq, context);
+    }
+
+    public static Query literalMatchesAllArrayRef(Literal<?> probe, Reference candidates, LuceneQueryBuilder.Context context) {
+        // 1 = ALL(array_col)
+        // --> 1 = array_col[1] and 1 = array_col[2] and 1 = array_col[3]
+        // --> not(1 != array_col[1] or 1 != array_col[2] or 1 != array_col[3])
+        // --> not(1 != ANY(array_col))
+        return new BooleanQuery.Builder()
+            .add(Queries.not(AnyNeqOperator.literalMatchesAnyArrayRef(probe, candidates)), BooleanClause.Occur.MUST)
+            .add(IsNullPredicate.refExistsQuery(candidates, context, false), BooleanClause.Occur.FILTER)
+            .build();
+    }
+
+    @Override
+    public @Nullable Query toQuery(Function function, LuceneQueryBuilder.Context context) {
+        List<Symbol> args = function.arguments();
+        Symbol probe = args.get(0);
+        Symbol candidates = args.get(1);
+        while (candidates instanceof Function fn && fn.signature().equals(ArrayUnnestFunction.SIGNATURE)) {
+            candidates = fn.arguments().get(0);
+        }
```

Comment:
I think this deserves some explanation. `toQuery` implementations of scalars shouldn't know about any other functions like this.

Reply:
I still don't fully understand why this needs to be specific for unnest. Couldn't this unwrap whatever is in candidates until it gets either the array function or a literal/reference?

Reply:
Sorry I don't understand.. `candidates.valueType` is array type and we are exceptionally unwrapping `ArrayUnnestFunction` for levelling the array dimensions. I am not sure what else could be unwrapped here.